ehr_ml/clmbr/featurizer.py

```diff
@@ -0,0 +1,241 @@
+                code_dropout=config["code_dropout"],
+        ) as batches:
+            for i, batch in enumerate(batches):
+                values, non_text_loss = self.model(batch)
+                del values
+
+                self.optimizer.zero_grad()
+                non_text_loss.backward()
+                self.optimizer.step()
+
+                del non_text_loss
+                del batch
+                if pbar is not None:
+                    pbar.update(1)
+                elif i % 2000 == 0:
+                    logging.info(f"Seen batch {i}")
+    
+    def fit(self, dataset, use_pbar = True):
+        self.model.train()
+        
+        model_dir = self.model.config["model_dir"]
+        num_epochs = self.model.config["epochs_per_cycle"]
+        
+        self.optimizer = self._build_adam_optimizer(dataset)
+
+        checkpoint_dir = os.path.join(model_dir, 'checkpoints')
+        os.makedirs(checkpoint_dir, exist_ok=True)
+        
+        best_val_loss = None
+        best_val_loss_epoch = None
+
+        pbar = tqdm(total = self.optimizer.defaults['t_total']) if use_pbar \
+                            else None
+        loss_file = open(os.path.join(model_dir, "losses"), "w")
+
+        logging.info("Start training")
+        for epoch in range(num_epochs):
+            logging.info("About to start epoch %s", epoch)
+            if pbar is not None:
+                pbar.set_description(f"Epoch {epoch}")
+            self._train_epoch(dataset, pbar = pbar)
+            logging.info("Epoch %s is complete", epoch)
+
+            if pbar is not None:
+                pbar.set_description(f"Evaluating epoch {epoch}")
+            train_loss = self.evaluate(dataset, is_val=False, num_batches = 2000)
+            val_loss = self.evaluate(dataset, is_val=True, num_batches = 2000)
+
+            logging.info("Train loss: %s", train_loss)
+            logging.info("Val loss: %s", val_loss)
+
+            loss_file.write("Epoch {}\n".format(epoch))
+            loss_file.write("Train loss {}\n".format(train_loss))
+            loss_file.write("Val loss {}\n".format(val_loss))
+            loss_file.write("\n")
+            loss_file.flush()
+
+            if best_val_loss is None or val_loss < best_val_loss:
+                best_val_loss = val_loss
+                best_val_loss_epoch = epoch
+
+                if os.path.exists(os.path.join(model_dir, "best")):
+                    os.unlink(os.path.join(model_dir, "best"))
+
+                torch.save(self.model.state_dict(), os.path.join(model_dir, "best"))
+                logging.info("Saving best model to %s", os.path.join(model_dir, "best"))
+
+        if pbar is not None:
+            pbar.close()
+        loss_file.close()
+        logging.info("Training complete!")
+        
+    def evaluate(self,
+                 dataset: StrideDataset,
+                 is_val: bool = True,
+                 num_batches: int = None):
```

Comment:
Did you run the typechecker on this? int is an invalid type for this variable. Should be Optional[int] = None